### ISSUE TITLE
Render hook warnings

### ISSUE THREAD

**Issue by mvdstam** [brillout/vite-plugin-ssr#966]
### Description

Hi @brillout,

In our application, we get the following logs:

```
[vite-plugin-ssr@0.4.131][Warning] The render() hook defined by /renderer/_default.page.server.tsx > `export { render }` is taking more than 4 seconds
[vite-plugin-ssr@0.4.131][Warning] The render() hook defined by /renderer/_default.page.server.tsx > `export { render }` is taking more than 4 seconds
[vite-plugin-ssr@0.4.131][Warning] The render() hook defined by /renderer/_default.page.server.tsx > `export { render }` is taking more than 4 seconds
Error: [vite-plugin-ssr@0.4.131] Hook timeout: the render() hook defined by /renderer/_default.page.server.tsx > `export { render }` didn't finish after 40 seconds
    at Timeout._onTimeout (/app/node_modules/vite-plugin-ssr/dist/cjs/utils/callHookWithTimeout.js:29:34)
Error: [vite-plugin-ssr@0.4.131] Hook timeout: the render() hook defined by /renderer/_default.page.server.tsx > `export { render }` didn't finish after 40 seconds
    at Timeout._onTimeout (/app/node_modules/vite-plugin-ssr/dist/cjs/utils/callHookWithTimeout.js:29:34)
Error: [vite-plugin-ssr@0.4.131] Hook timeout: the render() hook defined by /renderer/_default.page.server.tsx > `export { render }` didn't finish after 40 seconds
    at Timeout._onTimeout (/app/node_modules/vite-plugin-ssr/dist/cjs/utils/callHookWithTimeout.js:29:34)
```

The requests, however, don't take 4 seconds at all and seem to render just fine. The `render` hook of our application looks like this:

```ts
async function render(pageContext: PageContextServer) {
  const Layout = (pageContext.exports.Layout || Default) as ({ children }: { children: ReactNode }) => React.JSX.Element
  const { Page, pageProps } = pageContext
  const stylesServer = createStylesServer(StylesCache)
  const apolloClient = makeApolloClient(true, pageContext.cookie)
  const helmetContext = {}
  const { getState, setState } = useStore
  const log = getState().log.getSubLogger({ prefix: ['_default.page.server', 'render'] })

  if (!Page) throw new Error('My render() hook expects pageContext.Page to be defined')

  setState((state) => ({ ...state, apollo: apolloClient }))

  log.debug('Initializing state')
  await getState().initialize()
  log.debug('State initialized')

  log.debug('Rendering application')
  // Renders the application tree, with all use*Query Apollo hooks being resolved recursively
  const pageHtml = await renderToStringWithData(
    <React.StrictMode>
      <GoogleReCaptchaProvider reCaptchaKey={import.meta.env.VITE_RECAPTCHA_KEY}>
        <PageShell {...{ pageContext, helmetContext, apolloClient }}>
          <Layout>
            <Page {...pageProps} />
          </Layout>
        </PageShell>
      </GoogleReCaptchaProvider>
    </React.StrictMode>
  )

  log.debug('Application rendered')

  // @ts-ignore
  const { helmet } = helmetContext

  // noinspection HtmlRequiredTitleElement
  const documentHtml = escapeInject`<!DOCTYPE html>
<html lang="nl" class='min-h-full'>
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1,minimum-scale=1,user-scalable=0">
    <link rel="stylesheet" href="https://use.typekit.net/dbp5xgb.css" />

    ${dangerouslySkipEscape(helmet.title.toString())}
    ${dangerouslySkipEscape(helmet.meta.toString())}
    ${dangerouslySkipEscape(helmet.link.toString())}

    <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
    <link rel="manifest" href="/site.webmanifest">
    <link rel="mask-icon" href="/safari-pinned-tab.svg" color="#f7f8fc">
    <meta name="msapplication-TileColor" content="#f7f8fc">
    <meta name="theme-color" content="#f7f8fc">
    <meta name="apple-mobile-web-app-status-bar-style" content="default">
    <meta name="apple-mobile-web-app-capable" content="yes">

    ${dangerouslySkipEscape(renderToStaticMarkup(<ServerStyles html={pageHtml} server={stylesServer} />))}
  </head>
  <body class="min-h-screen flex flex-col text-brand-text-dark">
    <div id="react-root" class='flex-1 flex flex-col'>${dangerouslySkipEscape(pageHtml)}</div>
  </body>
</html>
`

  return {
    documentHtml,
    pageContext: {
      apolloState: apolloClient.extract(),
      zustandState: getState().serialize(),
    },
  }
}
```

The `await getState().initialize()` runs graphql queries (using Apollo) to initialize the application, after which the application is actually rendered. When rendering the application, more Apollo GraphQL queries may be executed under the hood depending on the state (ie user is logged in) after initialization. The issue only seems to happen when a user is logged in, where more queries are executed/resolved during rendering since more data may need to be fetched. The rendered final result seems to be correct.

Hope you can point me in the right direction. Thanks in advance for your time.

**Comment by brillout** [comment]
> don't take 4 seconds at all

I doubt it: are you sure your hooks are running below the thresholds? If you're confident they do, then create a minimal reproduction and I'll have a look at it.

Closing since I suspect that it isn't a bug.

We can add a feature to increase the warning thresholds, in case that's something you need.

**Comment by mvdstam** [comment]
The HTTP requests absolutely don’t take 4 seconds to complete, let alone 40 seconds. What I’m seeing, is these logs appearing _after_ the HTTP requests are completed and not _during_.

Is it possible that there is some kind of process/callback that doesn’t get resolved completely internally? I’m not seeing the same issues clientside. I have zero experience with SSR prior to my current project, so perhaps I’m simply overseeing something obvious causing this behavior.

Verstuurd vanaf mijn iPhone

Op 21 jun. 2023 om 18:00 heeft Rom ***@***.***> het volgende geschreven:

﻿

don't take 4 seconds at all

I doubt it: are you sure your hooks are running below the thresholds? If you're confident they do, then create a minimal reproduction and I'll have a look at it.

Closing since I suspect that it isn't a bug.

We can add a feature to increase the warning thresholds, in case that's something you need.

**Comment by brillout** [comment]
Have a look at `node_modules/vite-plugin-ssr/dist/cjs/utils/executeHook.js`: as you can see the timeouts are cleared once the promise is resolved.

**Comment by mvdstam** [comment]
I know and I believe you, but I can assure you that it doesn’t take 4 seconds on my end to see a rendered page, let alone 40 seconds causing the timeout messages. And yet, something is causing asynchronous processes to keep running _after_ Express has sent a response back to the client, causing the entire container running the application to clog up.

Since it’s (reasonably) reproducible, I’m going to take a fresh look at this tomorrow and come back to you on this as soon as possible.

Verstuurd vanaf mijn iPhone

Op 21 jun. 2023 om 20:44 heeft Rom ***@***.***> het volgende geschreven:

﻿

Have a look at node_modules/vite-plugin-ssr/dist/cjs/utils/executeHook.js: as you can see the timeouts are cleared once the promise is resolved.

**Comment by brillout** [comment]
👍 Looking forward to it.

**Comment by mvdstam** [comment]
Hi @brillout,

I figured it out, and it's definitely not a bug like you said.

I made a huge mistake in my `_default.page.server.tsx`, which re-used my [Zustand](https://github.com/pmndrs/zustand) store for every request that the server handled. I completely forgot the fact that NodeJS spawns a running server that keeps running as a process, in contrast to for example a PHP server that runs one process per request. Since the Zustand Store also did some asynchronous initialization (dependening on whether or not you're logged in), things went haywire pretty quick. So although the HTTP request handled by the Express loop returned pretty quickly, internally all kinds of processes kept running because the original state store never got cleaned up.

In other words, mystery solved and I can finally sleep again.

Cheers!
